src/compiler/checker.ts

```diff
@@ -8359,7 +8359,7 @@ namespace ts {
         }
 
         // This function assumes the constituent type list is sorted and deduplicated.
-        function getUnionTypeFromSortedList(types: Type[], aliasSymbol?: Symbol, aliasTypeArguments?: Type[]): Type {
+        function getUnionTypeFromSortedList(types: Type[], unionOfUnitTypes: TypeFlags, aliasSymbol?: Symbol, aliasTypeArguments?: Type[]): Type {

```

Comment:
Possible optimization: `unionOfUnitTypes` should just be `additionalFlags`, and we should calculate `getPropagatingFlagsOfTypes` piece-wise while doing `includes`, and just pass them in (rather then making _another_ pass over the type here to calculate them).... although I know right now to save flag space we re-purpose the propagating flags during include calculation, which could make that.... hard.

Reply:
Yeah, but I don't really want to mess with that now.